source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,565 @@
+    case SpvOp::SpvOpBitwiseXor:
+      assert(operand_words.size() == 2);
+      return operand_words.front() ^ operand_words.back();
+
+    // Logical
+    case SpvOp::SpvOpLogicalEqual:
+      assert(operand_words.size() == 2);
+      return ((bool)operand_words.front()) == ((bool)operand_words.back());
+    case SpvOp::SpvOpLogicalNotEqual:
+      assert(operand_words.size() == 2);
+      return ((bool)operand_words.front()) != ((bool)operand_words.back());
+    case SpvOp::SpvOpLogicalOr:
+      assert(operand_words.size() == 2);
+      return ((bool)operand_words.front()) || ((bool)operand_words.back());
+    case SpvOp::SpvOpLogicalAnd:
+      assert(operand_words.size() == 2);
+      return ((bool)operand_words.front()) && ((bool)operand_words.back());
+    case SpvOp::SpvOpLogicalNot:
+      assert(operand_words.size() == 1);
+      return !(bool)operand_words.front();
+    case SpvOp::SpvOpSelect:
+      assert(operand_words.size() == 3);
+      return ((bool)operand_words.front()) ? operand_words.at(1)
+                                           : operand_words.at(2);
+
+    // Comparison
+    case SpvOp::SpvOpIEqual:
+      assert(operand_words.size() == 2);
+      return operand_words.front() == operand_words.back();
+    case SpvOp::SpvOpINotEqual:
+      assert(operand_words.size() == 2);
+      return operand_words.front() != operand_words.back();
+    case SpvOp::SpvOpULessThan:
+      assert(operand_words.size() == 2);
+      return operand_words.front() < operand_words.back();
+    case SpvOp::SpvOpSLessThan:
+      assert(operand_words.size() == 2);
+      return ((int32_t)operand_words.front()) < ((int32_t)operand_words.back());
+    case SpvOp::SpvOpUGreaterThan:
+      assert(operand_words.size() == 2);
+      return operand_words.front() > operand_words.back();
+    case SpvOp::SpvOpSGreaterThan:
+      assert(operand_words.size() == 2);
+      return ((int32_t)operand_words.front()) > ((int32_t)operand_words.back());
+    case SpvOp::SpvOpULessThanEqual:
+      assert(operand_words.size() == 2);
+      return operand_words.front() <= operand_words.back();
+    case SpvOp::SpvOpSLessThanEqual:
+      assert(operand_words.size() == 2);
+      return ((int32_t)operand_words.front()) <=
+             ((int32_t)operand_words.back());
+    case SpvOp::SpvOpUGreaterThanEqual:
+      assert(operand_words.size() == 2);
+      return operand_words.front() >= operand_words.back();
+    case SpvOp::SpvOpSGreaterThanEqual:
+      assert(operand_words.size() == 2);
+      return ((int32_t)operand_words.front()) >=
+             ((int32_t)operand_words.back());
+    default:
+      assert(false);
+      return 0;
+  }
+}
+
+// A helper function to check the validity for broadcast operations. Returns
+// true if all of the given operands are:
+//  1) vector type front-end constants;
+//  2) same in the number of components.
+// This method is only used in assertions for now.
+bool AllInOperandValuesAreConstVectorsInSameLength(
```

Comment:
Wow, this function name is a little bit long. (You can get some kind of camel case badge I think.) ;)

What about `HasHomogeneousOperands` or something similar?